iceoryx_hoofs/include/iceoryx_hoofs/cxx/prefix_tree.hpp

```diff
@@ -0,0 +1,290 @@
+///        search result set, which can contain at most Capacity values (worst-case).
+///        The search interface avoids copying for efficiency and allows direct
+///        access of the found values (similar to a cxx::vector).
+/// @tparam Value type of values to be stored
+/// @tparam Capacity maximum number of values to be stored.
+/// @tparam MaxKeyLength maximum length of the key strings
+/// @note relocatable, i.e. logical state can be copied using memcpy
+template <typename Value, uint32_t Capacity, uint32_t MaxKeyLength = 128>
+class PrefixTree
+{
+  public:
+    // We only support specific string keys for now.
+    // NB: In general any sequence of letters with an order would work
+    //     but currently we only support char letters.
+    using Key = cxx::string<MaxKeyLength>;
+
+  public:
+    /// @brief Construct an empty prefix tree.
+    PrefixTree() noexcept;
+
+    ~PrefixTree() noexcept;
+
+    /// @todo implement copy later if needed, default may suffice since it is relocatable
+    ///       move is not intended (would be a copy)
+    PrefixTree(const PrefixTree& other) = delete;
+    PrefixTree(PrefixTree&& other) = delete;
+    PrefixTree& operator=(const PrefixTree&) = delete;
+    PrefixTree& operator=(PrefixTree&&) = delete;
+
+    /// @brief Insert a key and a correspoding value into the tree.
+    /// @param key key to be inserted
+    /// @param value value to be inserted
+    /// @return true if insertion was successful, false otherwise (capacity exceeded)
+    /// @note allows multiple values associated with a key, including duplicate values
+    bool insert(const Key& key, const Value& value) noexcept;
+
+    /// @brief Remove a key and all its associated values from the tree.
+    /// @param key key to be removed
+    /// @return true if was removed, false otherwise (key did not exist)
+    bool remove(const Key& key) noexcept;
+
+    /// @brief Remove a key and a corresponding value from the tree.
+    /// @param key key to be removed
+    /// @param value value to be removed
+    /// @return true if (key, value) was removed, false otherwise
+    ///         (key with this value did not exist)
+    bool remove(const Key& key, const Value& value) noexcept;
+
+    /// @note: pointer interface (Value*) is required for efficiency
+    ///        and to access values for modification
+
+    /// @todo value versions of find and similar functions? - copies values out
+    /// @todo callback interface of find and similar functions?
+    ///       pass a lambda to be invoked on matches to avoid returning large capacity
+    ///       containers
+
+    /// @brief Find all values corresponding to a given key.
+    /// @param key key whose values we want to find
+    /// @return pointers to values matching the key
+    /// @note returned pointers allow modification of corresponding values in the tree
+    cxx::vector<Value*, Capacity> find(const Key& key) const noexcept;
```

Comment:
Interface is suitable for intended use cases but can of course be extended. Returning values can be done in addition, but should not be the default.

Since it has multi-map semantics (required for our use case), I opted against `operator[]` for value access (there can be more values per key and we need to support this).